posthog/warehouse/models/modeling.py

```diff
@@ -0,0 +1,506 @@
+
+    def __init__(self, parent: str, query: str):
+        super().__init__(
+            f"The parent name {parent} does not correspond to an existing PostHog table, Data Warehouse Table, or Data Warehouse Saved Query."
+        )
+        self.query = query
+
+
+class ModelPathAlreadyExistsError(Exception):
+    """Exception raised when trying to create paths for a model that already has some."""
+
+    def __init__(self, model_name: str):
+        super().__init__(f"Model {model_name} cannot be created as it already exists")
+
+
+class ModelPathDoesNotExistError(Exception):
+    """Exception raised when trying to update paths for a model that doesn't exist."""
+
+    def __init__(self, model_name: str):
+        super().__init__(f"Model {model_name} doesn't exist")
+
+
+class DataWarehouseModelPathManager(models.Manager["DataWarehouseModelPath"]):
+    """A model manager that implements some common path operations."""
+
+    def create_from_saved_query(self, saved_query: DataWarehouseSavedQuery) -> "list[DataWarehouseModelPath]":
+        """Create one or more model paths from a new `DataWarehouseSavedQuery`.
+
+        Creating one or more model paths from a new `DataWarehouseSavedQuery` is straight-forward as we
+        don't have to worry about this model having its own children paths that need updating: We are
+        only adding a leaf node to all parents' paths. We check this model indeed does not exist to
+        ensure that is the case.
+
+        Raises:
+            ValueError: If no paths exists for the provided `DataWarehouseSavedQuery`.
+        """
+        if self.filter(team=saved_query.team, saved_query=saved_query).exists():
+            raise ModelPathAlreadyExistsError(saved_query.name)
+
+        return self.create_leaf_paths_from_query(
+            query=saved_query.query["query"],
+            team=saved_query.team,
+            created_by=saved_query.created_by,
+            label=saved_query.id.hex,
+            saved_query_id=saved_query.id,
+        )
+
+    def create_leaf_paths_from_query(
+        self,
+        query: str,
+        team: Team,
+        label: str,
+        created_by: User | None = None,
+        saved_query_id: uuid.UUID | None = None,
+        table_id: uuid.UUID | None = None,
+    ) -> "list[DataWarehouseModelPath]":
+        """Create all paths to a new leaf model.
+
+        A path will be created for each parent, as extracted from the given query.
+        """
+        base_params = {
+            "team": team,
+            "created_by": created_by,
+            "saved_query_id": saved_query_id,
+            "table_id": table_id,
+        }
+
+        with transaction.atomic():
+            parent_paths = self.get_or_create_query_parent_paths(query, team=team)
+
+            results = self.bulk_create(
+                [
+                    DataWarehouseModelPath(id=uuid7(), path=[*model_path.path, label], **base_params)
```

Comment:
Cycles should not be possible here, but it did make me realize there are a couple of bugs that can cause them:
* We check that the saved query path doesn't exist already, so it must be new. However, this check happens outside the transaction block, so there could be some problems during high-concurrency. The solution is to move the check inside the transaction block.
* Creating a cycle only with CREATE operations requires a query that selects from a parent that doesn't yet exist: There must be a node that selects from a node that will be its descendant to form a cycle, but since nodes are created in a serializable fashion, necessarily one is created first while the other doesn't exist. We check that all parents have paths created to prevent nodes that select from non-existent parents, but actually we weren't properly checking that at least one path is returned, so empty paths could work instead of raising a `UnknownParentError`. The solution is to add a `if not <paths>: raise UnknownParentError`.

Reply:
I've added a test to check for the second bug, the first one may be too hard to test as it involves concurrency...